oak_functions/client/java/src/AeadEncryptor.java

```diff
@@ -1,39 +1,92 @@
-import com.google.crypto.tink.subtle.AesGcmJce;
+import com.google.oak.functions.client.Util;
+import java.lang.IllegalArgumentException;
 import java.security.GeneralSecurityException;
+import java.security.SecureRandom;
+import java.util.Arrays;
+import javax.crypto.Cipher;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.GCMParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
 
 /**
  * Implementation of Authenticated Encryption with Associated Data (AEAD).
  * https://datatracker.ietf.org/doc/html/rfc5116
  */
 public class AeadEncryptor {
-    private AesGcmJce encryptor;
+    private static final String ENCRYPTION_ALGORITHM = "AES";
+    private static final String AEAD_ALGORITHM = "AES/GCM/NoPadding";
+    private static final int KEY_LENGTH_BITS = 256;
+    private static final int TAG_LENGTH_BITS = 128;
+    private static final int INITIALIZATION_VECTOR_LENGTH_BYTES = 12;
 
-    public AeadEncryptor(byte[] key) throws GeneralSecurityException {
-        // TODO(#2120): Use a wrapper AEAD interface and don't use `tink.subtle`.
-        this.encryptor = new AesGcmJce(key);
+    private SecretKey key;
+    private SecretKeySpec keySpecification;
+
+    public AeadEncryptor(byte[] session_key) throws IllegalArgumentException {
+        if (session_key.length * 8 != KEY_LENGTH_BITS) {
+            throw new IllegalArgumentException(
+                String.format("Incorrect key length: %d, expected %d", session_key.length * 8, KEY_LENGTH_BITS)
+            );
+        }
+        key = new SecretKeySpec(session_key, 0, session_key.length, AEAD_ALGORITHM);
+        keySpecification = new SecretKeySpec(key.getEncoded(), ENCRYPTION_ALGORITHM);
     }
 
     /**
-     * Encrypts `data` using `AeadEncryptor::encryptor`.
-     * The resulting encrypted data is prefixed with a random 12 bit nonce.
+     * Encrypts `data` using AES-GCM.
+     * The resulting encrypted data is prefixed with a random 12 bit initialization vector.
      */
     public byte[] encrypt(byte[] data) throws GeneralSecurityException {
+        Cipher encryptor = Cipher.getInstance(AEAD_ALGORITHM);
+
+        // Create a random initialization vector.
+        byte[] initializationVector = generateInitializationVector(INITIALIZATION_VECTOR_LENGTH_BYTES);
+        GCMParameterSpec gcmParameterSpecification = new GCMParameterSpec(TAG_LENGTH_BITS, initializationVector);
+
+        // Initialize encryptor.
+        encryptor.init(Cipher.ENCRYPT_MODE, keySpecification, gcmParameterSpecification);
+
         // Additional authenticated data is not required for the remotely attested channel,
         // since after session key is established client and server exchange messages with a
         // single encrypted field.
-        return this.encryptor.encrypt(data, null);
+        byte[] encryptedData = encryptor.doFinal(data);
+
+        // Add `initializationVector` as a prefix to the `encryptedData`.
+        return Util.concatenate(initializationVector, encryptedData);
```

Comment:
Alternatively, we could just return an object with the IV and the payload as separate fields, and we could serialize these as protobuf on the wire for Oak, instead of just concatenating them together, which effectively just a custom serialization anyways. e.g. you could have an EncryptedValue POJO for this with those two fields, and decrypt could just accept that as input too.